src/Fmt_ast.ml

```diff
@@ -1730,8 +1781,400 @@ and fmt_expression c ?(box= true) ?epi ?eol ?parens ?ext
+                 ) )
+           $ fmt_if Poly.(fields <> []) "@\n"
+           $ hvbox 0
+               (list fields "@\n" (fun cf -> fmt_class_type_field c ctx cf))
+           )
+       $ fmt_or_k Poly.(fields <> []) (fmt "@\n") (fmt "@ ")
+       $ fmt "end" ))
+
+and fmt_class_type c ?(box= true) ({ast= typ} as xtyp) =
+  protect (Cty typ)
+  @@
+  let {pcty_desc; pcty_loc; pcty_attributes} = typ in
+  let doc, atrs = doc_atrs pcty_attributes in
+  Cmts.fmt c.cmts pcty_loc
+  @@ ( if List.is_empty atrs then Fn.id
+     else fun k -> k $ fmt_attributes c ~key:"@" atrs (fmt "") )
+  @@
+  let parens = parenze_cty xtyp in
+  ( hvbox_if box 0 @@ wrap_if parens "(" ")"
+  @@
+  let ctx = Cty typ in
+  match pcty_desc with
+  | Pcty_constr ({txt; loc}, params) ->
+      let params = List.map params ~f:(fun x -> (x, Invariant)) in
+      Cmts.fmt c.cmts loc @@ fmt_class_params c ctx ~epi:(fmt "@ ") params
+      $ fmt_longident txt
+  | Pcty_signature {pcsig_self; pcsig_fields} ->
+      fmt_class_signature c ~ctx ~parens pcsig_self pcsig_fields
+  | Pcty_arrow (_, _, _) ->
+      let arg_label lbl =
+        match lbl with
+        | Nolabel -> fmt ""
+        | Labelled l -> str l $ fmt ":"
+        | Optional l -> fmt "?" $ str l $ fmt ":"
+      in
+      let xt1N = sugar_class_arrow_typ c (sub_cty ~ctx typ) in
+      hvbox_if box 0
+        (list xt1N "@;-> " (fun (lI, xtI) ->
+             hvbox 0
+               ( arg_label lI
+               $
+               match xtI with
+               | `core_type ct -> fmt_core_type c ct
+               | `class_type ct -> fmt_class_type c ct ) ))
+  | Pcty_extension ext -> fmt_extension c ctx "%" ext
+  | Pcty_open (flag, {txt}, cl) ->
+      hvbox 0
+        ( str "let open"
+        $ fmt_if Poly.(flag = Override) "!"
+        $ str " " $ fmt_longident txt $ fmt " in@;<1000 0>"
+        $ fmt_class_type c (sub_cty ~ctx cl) ) )
+  $ fmt_docstring c ~pro:(fmt "@ ") doc
+
+and fmt_class_expr c ?eol ?(box= true) ({ast= exp} as xexp) =
+  protect (Cl exp)
+  @@
+  let {pcl_desc; pcl_loc; pcl_attributes} = exp in
+  let parens = parenze_cl xexp in
+  let fmt_label lbl sep =
+    match lbl with
+    | Nolabel -> fmt ""
+    | Labelled l -> fmt "~" $ str l $ fmt sep
+    | Optional l -> fmt "?" $ str l $ fmt sep
+  in
+  let fmt_label_arg ?(box= box) ?epi ?parens (lbl, ({ast= arg} as xarg)) =
+    match (lbl, arg.pexp_desc) with
+    | (Labelled l | Optional l), Pexp_ident {txt= Lident i; loc}
+      when String.equal l i ->
+        Cmts.fmt c.cmts loc
+        @@ Cmts.fmt c.cmts ?eol arg.pexp_loc
+        @@ fmt_label lbl ""
+    | _ ->
+        hvbox_if box 2
+          (fmt_label lbl ":@," $ fmt_expression c ~box ?epi ?parens xarg)
+  in
+  let ctx = Cl exp in
+  let width xe =
+    String.length
+      (Format.asprintf "%t" (Cmts.preserve (fun () -> fmt_expression c xe)))
+  in
+  let fmt_args_grouped e0 a1N =
+    (* TODO: consider [e0] when groupping *)
```

Comment:
nit: s/groupping/grouping .

This is nontrivial due to the different types of e0 and a1N, right? Any other reason?

Reply:
no other reason.